cmd/cstor-pool-mgmt/volumereplica/volumereplica.go

```diff
@@ -669,3 +671,172 @@ func GetAndUpdateReplicaID(cvr *apis.CStorVolumeReplica) error {
 	}
 	return nil
 }
+
+// GetAndUpdateSnapshotInfo get the snapshot list from ZFS and updates in CVR status.
+// Execution happens in following steps:
+// 1. Get snapshot list from ZFS
+// 2. Checks whether above snapshots exist on CVR under Status.Snapshots:
+//    2.1 If snapshot doesn't exist then get the info of snapshot from ZFS and update
+//        the details in CVR.Status.Snapshots
+// 3. Verify and delete the snapshot details on CVR if it is deleted from ZFS
+// 4. Update the pending list of snapshots by verifying with snapshot list obtained from step1
+// 5. If replica is under rebuilding get the snapshot list from peer CVR and update them
+//    under pending snapshot list
+func GetAndUpdateSnapshotInfo(
+	clientset clientset.Interface, cvr *apis.CStorVolumeReplica) error {
+	volName := cvr.GetLabels()[string(apis.PersistentVolumeCPK)]
+	dsName := PoolNameFromCVR(cvr) + "/" + volName
+
+	snapList, err := GetSnapshotList(dsName)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get the list of snapshots")
+	}
+
+	// Add/Delete a snapshot if it doesn't exist in CVR but exists in ZFS
+	err = addOrDeleteSnapshotListInfo(cvr, snapList)
+	if err != nil {
+		return errors.Wrapf(err, "failed to add snapshot list info")
+	}
+
+	// If CVR is in rebuilding go and get snapshot information
+	// from other replicas
+	//if cvr.Status.Phase == apis.CVRStatusRebuilding {
+	//	err = getAndUpdatePendingSnapshotsList(clientset, cvr)
+	//}
+	return nil
+}
+
+//func getAndUpdatePendingSnapshotsList(
+//	clientset clientset.Interface, cvr *apis.CStorVolumeReplica) error {
+//	peerCVRList, err := getPeerReplicas(clientset, cvr)
+//	if err != nil {
+//		return errors.Wrapf(err, "failed to get peer CVRs info of volume %s", volName)
+//	}
+// TODO: Perform following steps
+// 1. Get the snapshot info from peer replicas
+// 2. If snapshot info doesn't exist under CVR.Status.Snapshots then
+// add it under CVR.Status.PendingSnapshots
+//}
+
+// addOrDeleteSnapshotListInfo adds/deletes the snapshots in CVR
+// It performs below steps:
+// 1. Add snapshot if it doesn't exist in CVR but exist on ZFS.
+// 2. Delete snapshot if exist in CVR but not in ZFS.
+// 3. Delete pending snapshots in CVR if exist in ZFS.
+// NOTE: Below function will get the snapshot info from ZFS
+func addOrDeleteSnapshotListInfo(
```

Comment:
in the case of non-errors, better to raise event or add log to tell what got added or deleted?

Reply:
We are already adding CR why we need it again.

Reply:
added